sdk/internal/perf/perf.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+package perf
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+	"text/tabwriter"
+	"time"
+
+	"github.com/spf13/pflag"
+)
+
+var (
+	debug     bool
+	Duration  int
+	TestProxy string
+	WarmUp    int
+	Parallel  int
+	wg        sync.WaitGroup
+)
+
+type PerfTest interface {
+	// GetMetadata returns the name of the test
+	GetMetadata() PerfTestOptions
+
+	GlobalSetup(context.Context) error
+
+	Setup(context.Context) error
+
+	Run(context.Context) error
+
+	Cleanup(context.Context) error
+
+	GlobalCleanup(context.Context) error
+}
+
+type HTTPClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type PerfTestOptions struct {
+	ParallelIndex int
+	ProxyInstance HTTPClient
+	Name          string
+}
+
+type NewPerfTest func(options *PerfTestOptions) PerfTest
+
+func runGlobalSetup(p PerfTest) error {
+	setRecordingMode("live")
+	err := p.GlobalSetup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func runSetup(p PerfTest) error {
+	err := p.Setup(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// runTest takes care of the semantics of running a single iteration. It returns the number of times the test ran as an int, the exact number
+// of seconds the test ran as a float64, and any errors.
+func runTest(p PerfTest, c chan runResult) {
+	defer wg.Done()
+
+	// If we are using the test proxy need to set up the in-memory recording.
+	if TestProxy != "" {
+		// First request goes through in Live mode
+		setRecordingMode("live")
+		err := p.Run(context.Background())
+		if err != nil {
+			c <- runResult{err: err}
+		}
+
+		// 2nd request goes through in Record mode
+		setRecordingMode("record")
+		err = start(p.GetMetadata().Name, nil)
+		if err != nil {
+			panic(err)
+		}
+		err = p.Run(context.Background())
+		if err != nil {
+			c <- runResult{err: err}
+		}
+		err = stop(p.GetMetadata().Name, nil)
+		if err != nil {
+			panic(err)
+		}
+
+		// All ensuing requests go through in Playback mode
+		setRecordingMode("playback")
+		err = start(p.GetMetadata().Name, nil)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	if WarmUp > 0 {
+		warmUpStart := time.Now()
+		for time.Since(warmUpStart).Seconds() < float64(WarmUp) {
+			err := p.Run(context.Background())
+			if err != nil {
+				c <- runResult{err: err}
+			}
+		}
+	}
+
+	timeStart := time.Now()
+	totalCount := 0
+	lastPrint := 1.0
+	perSecondCount := make([]int, 0)
+	w := tabwriter.NewWriter(os.Stdout, 16, 8, 1, ' ', tabwriter.AlignRight|tabwriter.Debug)
+	for time.Since(timeStart).Seconds() < float64(Duration) {
+		err := p.Run(context.Background())
+		if err != nil {
+			c <- runResult{err: err}
+		}
+		totalCount += 1
+
+		// Every second (roughly) we print out an update
```

Comment:
The per-second updates need to be aggregated across all the parallel instances, rather than printing one line per parallel instance.

Reply:
Fixed